### DISCUSSION TITLE
Split-screen to multiple monitors

### DISCUSSION THREAD

**Discussion by lurendrejer** [LizardByte/community#56] [Sunshine]
Hi, was just playing a split-screen game and thought to myself:

Would it be viable for Sunshine to present two 'outputs' ?
If one television could connect to the left or right side of the stream, split-screen multiplayer could be routed to different rooms.

Right now, we are often "forced" to by two copies of the same game to play local multiplayer.

**Comment by shodanx2** [comment]
I love this idea, the aspect ratio on the receiving end could be wonky.

This touches on a lot of thing I would like sunshine to do.

The easiest way to do what you ask is to send the full output to two moonlight clients

So how would that work, does sunshine maintain multiple independent TCP video streams, or does the sunshine client send the video as a single UDP broadcast or multicast over the network ?

How does multiple moonlight clients send controller input back to sunshine ? Maybe multiple client don't bother at all with controllers ?

As for each player seeing only their screen, maybe moonlight client could have a blackout option where the server or client can specify area coordinates that are to be kept black ? What about live-cropping the incoming video so it stretches to more of the video (losing some of the aspect ratio and some pixels like "pan and scan" or overscan video ?)


And the holy grail of this is, one gaming PC, streaming two full monitors (windows indirect displays, no physical monitors required) to two clients. One pc two gamers situation, this is certainly the future !

**Comment by lurendrejer** [comment]
I usually play splitscreen on the ultrawide - aspect ratio is not an issue in modern games.

Moonlight could implement this by having an option to only show part of the multicast stream.
It would just be a much cleaner implementation to have Sunshine present two seperate sessions of a 1920*2x1080 stream. 

When multicasting (i only tried this with parsec) input isn't a problem - controllers can be connected from multiple clients and works just fine.

It would probably just work if moonlight implemented a left/right 50% solution, actually. Having this serverside would require two sessions, two times the encryption, etc.

**Reply by lurendrejer** [reply]
https://github.com/moonlight-stream/moonlight-android/issues/1104

Posted in the moonlight-github instead.

**Comment by shodanx2** [comment]
If you can already make the game work in 3840x1080, for now you could try with just plain ffmpeg

like this

ffmpeg -f gdigrab -framerate 30 -video_size 1920x1080 -offset_x 0 -offset_y 0 -i desktop -vcodec hevc_nvenc -f mpegts srt://:6666?mode=listener
ffmpeg -f gdigrab -framerate 30 -video_size 1920x1080 -offset_x 1920 -offset_y 0 -i desktop -vcodec hevc_nvenc -f mpegts srt://:6667?mode=listener

On the receiving computers

ffplay -fs -f mpegts srt://yourserveriphere:6666?mode=caller
ffplay -fs -f mpegts srt://yourserveriphere:6667?mode=caller


However latency is probably too high, you need to reduce the buffers and use a directshow grabber instead of gdi grabber but I can't figure out how sunshine sets up ffmpeg to do this

**Reply by lurendrejer** [reply]
pairing what you posted, with this: https://sunshinestream.readthedocs.io/en/latest/about/advanced_usage.html#channels - might be a solution.

**Reply by lurendrejer** [reply]
it's bedtime in my end - it might be possible to just setup two channels, kill ffmpeg and start it again manually as you described.

**Comment by mirh** [comment]
> Right now, we are often "forced" to by two copies of the same game to play local multiplayer.

But are you, *actually*?
You seem to be describing the same effort [nucleus](https://nucleus-coop.github.io/) is supposed to take care of. 

Not sure why it should be up to the streaming server to handle this. 
Aren't games with built-in split-screen such as L4D2 or portal not already working out of the box?

**Comment by lurendrejer** [comment]
I did not know about nucleus, thank you. Nucleus solves a different problem though.

**Reply by mirh** [reply]
Yes, but.. I still have to understand what you are trying to propose here. 
If the client is connected to multiple monitors, this setup should get to be mirrored back to the host?
(and if yes, isn't "extending technology" such as eyefinity already enough for this use case?)

**Reply by shodanx2** [reply]
You would want sunshine to treat each splits of the screen as a different screen

Then have moonlight pick which screen to display

Then maybe install the amyuni indirect display driver ( https://www.amyuni.com/forum/viewtopic.php?t=3030 )
And create a single 1920x2160 display "indirect" display 

Maybe it could help to use fancyzone to subdivide the monitor into two monitor at the location of the in game split

This is a narrow use case but there is a broader use case here, multi monitor streaming

I have not tried it yet, but I suspect it's hard to run more than one instance of sunshine at a time on the same computer.

Imagine a simple setup with 3 physical display. In Sunshine you can choose one display to stream. In moonlight you choose one computer to stream from, as far as I know you can't change the streaming source monitor from moonlight

In an ideal world, Sunshine would let you, choose which of one or multiple monitor you want to stream. 
And you would run multiple moonlight clients, one for each screen.

Sunshine should understand "sub" monitors created by fancy zones, either as independant monitor, or as "logical monitors"  

In the backend, it's not clear how to do this. Do you screenscrab every requested monitor at once, use one hardware accelerated encoder to compress the entire thing and then server the stream to the network and let the moonlight client crop only what they want to see ?

Or, do you multiple screengrab process, each grabbing one monitor or sub monitor at a time, each using a separate hardware encoder process and moonlight only has to choose which monitor it wants to see ?

Here for example my setup
![image](https://user-images.githubusercontent.com/10621885/189507357-882b5d50-0308-441a-8f68-c94f54d7985e.png)

So the main monitor is connected directly with a hdmi cable,
The four side monitors are connected each to a Raspberry pi 4
The PC is running software called spacedesk that creates one indirect display (memory only display) and streams it as MJPEG over the network. 

If Sunshine could do multiple monitors, I could replace spacedesk with sunshine

And the OP could use the same thing to broadcast his split screen game to two separate monitors on the network

**Reply by mirh** [reply]
Good god, this is all over the place.. 
If running two copies of the same game on the host machine (i.e. nucleus's job) wasn't what OP was talking about, his idea was only having a stream of just a portion of the screen (or alternatively separate "watchable viewports" of the original screen).  
Assuming there are no problems with "very wide" resolutions streaming (at least until you hit the pixel limits of hardware encoding) and multicasting, the probably hardest part to implement here seems the UI. 

You basically want instead any-to-any monitor capabilities (including the host knowing and caring about the client topology). 
Which is pretty (if not completely) different from the gaming use case given that:

* games may often support ultrawide aspect ratios, but the concept of multi-monitor is foreign to 99.9% of them
* at least on Windows, it's absolute hell to control multiple input devices when they are not controllers (see all the work behind nucleus)

And this is not even the craziest part. Faking monitors to Windows is a [herculean task](https://superuser.com/a/1742744/567466) (yes took me three days to collect all that info). 
Like.. I'll grant to you that it got [easier](https://linustechtips.com/topic/1402610-creating-a-display-instance-without-a-monitor-or-a-dummy-plug/) in the last years, but taking care of that is basically worth its own separate project.

**Comment by lurendrejer** [comment]
It might just be a stupid midnight thought.

We were playing Gears5 on an ultrawide monitor, we had full 16:9 aspect ratio for each splitscreen player.
If sunshine could present monitor 1:Left and monitor 1:Right to moonlight, we could be playing a single instance of the game on two TV's (no matter what streaming client we used).

I ended up bying two copies of Gears, but i still think some split-screen games could benefit from that ability to have a TV for each player in seperate rooms.

**Reply by mirh** [reply]
So.. premising that multicasting should be already supported AFAIK, what you seem to be suggesting here is either:

* tinkering with the host to be able to only stream just some (and potentially multiple) portions of the screen
* having the [client](https://github.com/moonlight-stream/moonlight-android/issues/1104) choosing to display a part of the source

I wanted to immediately dismiss the later as something stupid/hacky/dirty (won't somebody think to the wasted network bandwidth?), but pondering better to the situation there may be cases where you want the server to have as low of an overhead as possible. Be it the one involved with cutting/encoding/packaging multiple (if smaller) streams, or even just wanting to use NVFBC.

**Reply by shodanx2** [reply]
Multicast would mean unlimited video and audio receivers on the network, for free.
Only the "master" client may tell the server about re-sending frames, changing stream settings etc.
But every other client can just watch and listen the streams, for free, no extra bandwidth needed.

I'm already doing this with a combo of ffmpeg and gstreamer, here are my command lines, I'm just about to wrap this up into a fully fledged application that also handles fake monitors (indirect display) management, and I'll finally have my any-to-any audio/video/control setup that I've always wanted and hoped sunshine would become.

video stream sender
`ffmpeg -hide_banner -f lavfi -i ddagrab=framerate=60:output_idx=3:video_size=1680x1050:offset_x=0:offset_y=0 -c:v h264_nvenc -preset llhp -tune ull -f mpegts udp://239.0.0.10:9998`
video stream receiver
`ffplay -hide_banner -fflags nobuffer -flags low_delay -probesize 20000 -analyzeduration 1 -strict experimental -framedrop  udp://239.0.0.10:9998`

audio stream sender
`gst-launch-1.0 -v wasapisrc loopback=true ! audioconvert ! udpsink host=239.0.0.2 port=9998`
audio stream listener
`gst-launch-1.0 -v udpsrc address=239.0.0.2 port=9998 multicast-group=239.0.0.1 caps="audio/x-raw,format=F32LE,rate=48000,channels=2" ! queue ! audioconvert ! autoaudiosink`

And you can have an enormous amount of stream even off a single puny gigabit lan connection, and this setup is far from optimized for optimal bandwidth (switching to h265 or AV1 to start with, but h264 means everything with silicon can listen in)

With barrier / inputleap you can easily switch which receiver has the actual control of the game (in my case the game runs in a VM in proxmox using GPU passthrough, I switch games by pausing the VM and switching to another VM seamlessly instead of stopping one game and starting another)

Really hoping Sunshine will better handle multi stream in the future. BTW I have found if the streaming region is continuous, you can stream multiple monitors at once, and then sub-crop the video stream on the receiving end. So @lurendrejer 's proposal is feasible using basic ffmpeg features and it only uses one video encoder to do it this way. 

I haven't re-tried sunshine/moonlight in quite a while, does the receiver work well in windowed mode nowadays and could it be made to work in passive, multicast receiver mode ?

**Reply by DavidSpivey** [reply]
Hey, is there any update on this? I am trying to multicast my screen, and your comments are the only things that remotely work, though it seems some of the command line options have changed in the latest ffmpeg.

**Reply by shodanx2** [reply]
I can tell you that in the sunshine documentation, when it talks about "multicast" it means multiple concurrent identical streams. 
That is two moonlight clients getting the same stream.
Apparently they each get their own encoder thread for bitrate variance but they can't watch different screen, I assume it's just because something is missing. 

But there is no way to do an "open loop" multicast (meaning, no bitrate scaling or any communication from a receiving client). It could be done, it's a core ffmpeg feature, but these controls are not broken out in sunshine.

**Reply by ReenigneArcher** [reply]
> they can't watch different screen, I assume it's just because something is missing.

Sunshine/Moonlight is not for watching screens. It's for streaming AND controlling them. The difficult part to solve would be sending input to separate applications.

**Reply by shodanx2** [reply]
@ReenigneArcher 

Certainly that would be great to have multiple client controlling multiple different applications.
But that is a whole other layer of complexity.
Just streaming multiple different monitor at once would be great.

The simple case to imagine is a dual monitor setup.

As far as I know moonlight can't stream two monitors with a single client.

And already there's two different use cases,

You could also have one laptop plus a tablet, each a separate moonlight client. Only the laptop cares about sending input back. 

Most base operating system do not have a concept of focusing more than one application, so trying to get multiple clients interacting with multiple applications on the same host, is simply too science-fiction for the current state of computing. 

Even having two mouse connected to one computer is basically Star Trek level technology. I think X11 has support for it, but wayland had a feature regression on that one. 

In my earlier case I was hoping for multicast, in which there is no client proper, just a transmission to the network with no designated receivers. The ffmpeg commands above permit this but not with the low latency of sunshine because the filter stack is too primitive compared with sunshine. 

Essentially there might be a stub moonlight client just triggering the sunshine server to transmit toward 239.0.0.1, then any number of receiver could display the signal. 

In my case I was using that to mirror my screens into the bedroom and kitchen, so I could keep an eye on things or keep watching my video when I was cooking and stuff.

As far as I'm concerned, moonlight is for conveniently streaming video with extreme low latency over the network. To and from any monitor, for any purpose and to stream user input controls in the other direction when possible. 

I've seen many other post that sunshine isn't for this or that, but Sunshine is VNC RDP SPICE pcanywhere bomgar teamviewer logmein chromecast and any number of app that all fundamentally do the same thing. 

Sunshine is just the best version of all of these and open source. It's so close to being everything one could want from any of these applications. 

Ah,  just had an idea for multiple client controlling multiple independent applications, 
On windows with https://github.com/stascorp/rdpwrap, I believe you could put each client into it's own "terminal service" so they'd essentially have each their own desktop, focus and mouse cursor.

**Reply by Hazer** [reply]
@shodanx2 I'm kinda torn on it.

Moonlight [IS a GameStreaming tool](https://moonlight-stream.org), meaning, interacting remotely with games, that's how it started. And Sunshine was born as a host for it.

But I do agree that it has potential outside of it, and it would be hypocrite of me, if I didn't agree, as I'm rarely using Moonlight/Sunshine for gaming, but using it daily as a productivity tool. And I also want extra flexibility on clients, streams and other aspects of Sunshine.

> Sunshine is VNC RDP SPICE pcanywhere bomgar teamviewer logmein chromecast and any number of app that all fundamentally do the same thing.

It's not, but I do understand why it feels like this, and I kinda use it like this too. A great open-source VNC project is RustDeck (if you don't know, maybe give it a try), but at the same time, it wasn't meeting my needs too, so here I am.

But some feature requests, sound to me are better suited as new moonlight clients, or even a potential new wave of clients that could be it's own thing. For example:

> I was hoping for multicast, in which there is no client proper, just a transmission to the network with no designated receivers
> [...]
> Essentially there might be a stub moonlight client just triggering the sunshine server to transmit toward 239.0.0.1, then any number of receiver could display the signal.

That would be a good proposition to a standalone "Something Multicast Client", maybe Moonlight-Multicast, maybe an enterely unrelated project, that supports connecting to Sunshine instance and other moonlight-protocol server instances, and multicast it using adequate protocols for multicasting with low bandwidth. But I don't think anything has to be done on Sunshine, or even Moonlight, for it to exist. It feels like a separated entity to me.

> The simple case to imagine is a dual monitor setup.
> As far as I know moonlight can't stream two monitors with a single client.
> You could also have one laptop plus a tablet, each a separate moonlight client. Only the laptop cares about sending input back.

I am currently using a setup where I primarily use my Sunshine host to work, where I created a virtual extended monitor, that I use Moonlight to see in another device, so that I can have multiple monitors working, and I'm using it as a productivity tool, not gaming. So, using a tool to create the virtual monitor, and adding another display, is kinda possible, but not possible to cast 2 monitors, into 2 other devices yet. With that said, I think devices like Apple Vision Pro may benefit of multi-screen streams too, so not only one to many clients, but many streams to a single client, may be useful, again, as a productivity tool.

So while I think some quality of life features may land on Moonlight, I have a feeling that productivity tools, maybe would be better with a new client for productivity that's not Moonlight, or at least, Moonlight having a new "Productivity Mode" where it's explicit to the users, that's not for gaming, and Sunshine may adapt to include support for it, but I don't know if the maintainers have interest on it, and someone has to start the new clients or proposing how a "Productivity Mode" would make sense to Moonlight.

**Comment by lurendrejer** [comment]
Multiple applications was never a thing.One application.Half the screen to each client.One client sends to controller 1And another client sends to controller 224. maj 2024 09.33 skrev shodanx2 ***@***.***>:
@ReenigneArcher
Certainly that would be great to have multiple client controlling multiple different applications.
But that is a whole other layer of complexity.
Just streaming multiple different monitor at once would be great.
The simple case to imagine is a dual monitor setup.
As far as I know moonlight can't stream two monitors with a single client.
And already there's two different use cases,
You could also have one laptop plus a tablet, each a separate moonlight client. Only the laptop cares about sending input back.
Most base operating system do not have a concept of focusing more than one application, so trying to get multiple clients interacting with multiple applications on the same host, is simply too science-fiction for the current state of computing.
Even having two mouse connected to one computer is basically Star Trek level technology. I think X11 has support for it, but wayland had a feature regression on that one.
In my earlier case I was hoping for multicast, in which there is no client proper, just a transmission to the network with no designated receivers. The ffmpeg commands above permit this but not with the low latency of sunshine because the filter stack is too primitive compared with sunshine.
Essentially there might be a stub moonlight client just triggering the sunshine server to transmit toward 239.0.0.1, then any number of receiver could display the signal.
In my case I was using that to mirror my screens into the bedroom and kitchen, so I could keep an eye on things or keep watching my video when I was cooking and stuff.
As far as I'm concerned, moonlight is for conveniently streaming video with extreme low latency over the network. To and from any monitor, for any purpose and to stream user input controls in the other direction when possible.
I've seen many other post that sunshine isn't for this or that, but Sunshine is VNC RDP SPICE pcanywhere bomgar teamviewer logmein chromecast and any number of app that all fundamentally do the same thing.
Sunshine is just the best version of all of these and open source. It's so close to being everything one could want from any of these applications.
Ah,  just had an idea for multiple client controlling multiple independent applications,
On windows with https://github.com/stascorp/rdpwrap, I believe you could put each client into it's own "terminal service" so they'd essentially have each their own desktop, focus and mouse cursor.

**Reply by mirh** [reply]
Multiple separate inputs to a single computer is pretty much a cakewalk on linux AFAIK
And even on windows.. you don't really need *mices*, as I already hinted above: https://github.com/SplitScreen-Me/splitscreenme-nucleus

> Multiple applications was never a thing.One application.Half the screen to each client.One client sends to controller 1And another client sends to controller

But if this was request, then the title has nothing to do with anything. 
You are just asking for multiple clients to be able to connect to the host.

**Reply by Hazer** [reply]
While there are this open-source project that does it, it by no means looks like a cakewalk endeavour to me, but I'd say that I think in that department, a possible interesting feature, to vote for support in Sunshine, would be instead of sending the input to the device, be able to assign a controller to an output channel, that other application could manually register to consume on the host and use as it pleases.

What I mean by that, this way, splitscreenme-nucleus that has the purpose of allowing for splitscreen on games that don't support it, could then listen to a "Sunshine-controller" from each connected client and use it accordingly. So Sunshine could launch it, provide the controller channel, and it would then do it's thing behind.

But if Sunshine ever creates a feature like that, [splitscreenme-nucleus](https://github.com/SplitScreen-Me/splitscreenme-nucleus), or another similar tool, would still need to add support to this new "Sunshine-controller interface", whatever that may look like. I don't know, at the moment, how much of an endeavour that would be to start at this point, but I think it's an interesting feature to allow for some external integration.

@lurendrejer 
> Right now, we are often "forced" to by two copies of the same game to play local multiplayer.
> Multiple applications was never a thing.One application.Half the screen to each client.One client sends to controller 1And another client sends to controller 2

The game already supports local split-screen on a single PC? Or we are talking about a game that don't support it? From your previous message, my understanding is that it doesn't support. If that's the case, you NEED 2 applications running, and for Windows, [splitscreenme-nucleus](https://github.com/SplitScreen-Me/splitscreenme-nucleus) purpose is to solve that, idk if there's something for linux. If you use Windows, at the moment, you could use VirtualHere to connect your controllers, then configure the controllers in nucleus, launch your game and then connect with moonlight clients to see the split-screen.

But if the game, already supports local split-screen, what I understand from your request, would be that each Moonlight Client, be recognized as a different controller input, and to stream only portions of the screen for each, right?

The game already supports local split-screen, or not? They are entirely different requests.

**Comment by lurendrejer** [comment]
The idea was to cast a split screen game to two separate clients.I might have been unclear 24. maj 2024 15.22 skrev mirh ***@***.***>:
Multiple separate inputs to a single computer is pretty much a cakewalk on linux AFAIK
And even on windows.. you don't really need mices, as I already hinted above: https://github.com/SplitScreen-Me/splitscreenme-nucleus

Multiple applications was never a thing.One application.Half the screen to each client.One client sends to controller 1And another client sends to controller

But if this was request, then the title has nothing to do with anything.
You are just asking for multiple clients to be able to connect to the host.

**Comment by lurendrejer** [comment]
The game supports splitscreen.I was only talking about local wide splitscreen streamed to two separate tv's.Modern split screen games support 2x 16:9 left/right 24. maj 2024 19.01 skrev Vithorio Polten ***@***.***>:
While there are this open-source project that does it, it by no means looks like a cakewalk endeavour to me, but I'd say that I think in that department, a possible interesting feature, to vote for support in Sunshine, would be instead of sending the input to the device, be able to assign a controller to an output channel, that other application could manually register to consume on the host and use as it pleases.
What I mean by that, this way, splitscreenme-nucleus that has the purpose of allowing for splitscreen on games that don't support it, could then listen to a "Sunshine-controller" from each connected client and use it accordingly. So Sunshine could launch it, provide the controller channel, and it would then do it's thing behind.
But if Sunshine ever creates a feature like that, splitscreenme-nucleus, or another similar tool, would still need to add support to this new "Sunshine-controller interface", whatever that may look like. I don't know, at the moment, how much of an endeavour that would be to start at this point, but I think it's an interesting feature to allow for some external integration.
@lurendrejer

Right now, we are often "forced" to by two copies of the same game to play local multiplayer.
Multiple applications was never a thing.One application.Half the screen to each client.One client sends to controller 1And another client sends to controller 2

The game already supports local split-screen on a single PC? Or we are talking about a game that don't support it? From your previous message, my understanding is that it doesn't support. If that's the case, you NEED 2 applications running, and for Windows, splitscreenme-nucleus purpose is to solve that, idk if there's something for linux. If you use Windows, at the moment, you could use VirtualHere to connect your controllers, then configure the controllers in nucleus, launch your game and then connect with moonlight clients to see the split-screen.
But if the game, already supports local split-screen, what I understand from your request, would be that each Moonlight Client, be recognized as a different controller input, right?
The game already supports local split-screen, or not? They are entirely different requests.

**Reply by Hazer** [reply]
Ok, got it. I added that to my research backlog.

**Comment by Hazer** [comment]
@ReenigneArcher I'm thinking that maybe we could split this split-screen discussion, into new smaller discussions, I'd propose the following discussion categories, and redirect people to each one for further comments, what do you think?

- Productivity Tools (Multi monitors/streams/multicast)
- ?Multicast (Maybe it's own discussion)
- Split game that support local multiplayer, to 2 Moonlight Clients
- Sunshine-controller interface/channel, so tools like [splitscreenme-nucleus](https://github.com/SplitScreen-Me/splitscreenme-nucleus) could listen to split-screen on non-local multiplayer games

**Reply by mirh** [reply]
Integration with nucleus seems like a killer feature, but putting aside this high-hanging fruit I'd say it's:
* just streaming selected parts/splits of the screen (which seems pretty duper trivial)
* multi-client (presumably this would be entirely transparent to them?)
* ? faking multimonitor support ([Virtual Display Driver](https://github.com/itsmikethetech/Virtual-Display-Driver) integration?)

**Comment by shodanx2** [comment]
Yes it should be split

For this discussion, split monitor gaming, I think it should happen in the moonlight client. 
ffmpeg has a crop filter, documented [here](https://video.stackexchange.com/a/4571/37266) that could do this.

The client would allow the user to set named crop zone  (main, player 1, player 2 etc..) and have a keyboard shortcut to switch between the zones.  I don't think the sunshine server needs to do anything to accomodate this.


As for multi monitor, it is already almost there.
The server can have multiple clients already.
Each client has their own encoder and separate stream. 
The only problem is that using the change monitor keyboard shortcut CTRL + ALT + SHIFT + F1/F12,  changes the capture monitor for all clients !
If that was resolved, most multi-monitor use cases would have a basic solution.
Probably no change would be needed client side. There would probably need to be a way to specify a default monitor per client (so left screen, is always left screen)

There is a workaround that appears to work but I have not yet tried it. Running multiple sunshine instances.
I managed to get multiple sunshine instance working but I didn't have a second client at the time of the test.
The first step for this solution, would be to make it easier to have multiple config files and multiple services created because right now you have to make the config files by hand, create the service etc.. It's a lot of work for a each user just to try  it !

For IP multicast, if the sunshine could be convinced to send the stream to 239.0.0.1 and just maintain a constant bitrate, that should be enough to allow multiple client to just playback the stream directly. Maybe ffplay could playback the stream itself.
